storage/innobase/fil/fil0crypt.cc

```diff
@@ -486,24 +561,34 @@ fil_space_crypt_t *fil_space_read_crypt_data(const page_size_t &page_size,
                     0) ==
       tablespace_key) {  // tablespace_key is all zeroes which means there is no
                          // tablepsace in mtr log
-    crypt_data->set_tablespace_key(NULL);
-    crypt_data->set_tablespace_iv(NULL);  // No tablespace_key => no iv
+    crypt_data->set_tablespace_key(nullptr);
   } else {
-    ut_ad(tablespace_key != NULL);
-    crypt_data->set_tablespace_key(tablespace_key);  // Since there is
-                                                     // tablespace_key present -
-                                                     // we also need to read
-                                                     // tablespace_iv
-    uchar tablespace_iv[ENCRYPTION_KEY_LEN];
-    ut_ad(tablespace_iv != NULL);
-    memcpy(tablespace_iv, page + offset + bytes_read, ENCRYPTION_KEY_LEN);
-    bytes_read += ENCRYPTION_KEY_LEN;
-    crypt_data->set_tablespace_iv(tablespace_iv);
+    crypt_data->set_tablespace_key(
+        tablespace_key);  // We are using the same iv for both
+                          // MK encryption and KEYRING encryption
   }
```

Comment:
This has been removed because it was redundant?

Reply:
Yes. This used to work like this - we have MK encrypted table with its iv (crypt_data->tablespace_iv). We start to re-encrypt this table with keyring encryption. We used to generate a new iv and put it into crypt_data->iv for keyring encryption. We used tablespace_iv when decrypting MK encrypted pages and crypt_data->iv when encrypting pages with Keyring. However, this crypt_data->iv can be the same as tablespace_iv. So I put tablespace_iv in crypt_data->iv and removed  tablespace_iv.

Reply:
no, on v2 server we abort the upgrade if there is a table in v1 format that is in a middle of MK to KEYRING rotation - i.e. it has tablespace_iv set.